thunder/distributed/prims.py

```diff
@@ -330,3 +389,74 @@ def synchronize_backward_rule(
         case _:
             utils.check(False, lambda: f"synchronize with unexpected {ddp_type=}")
     return synced_grad, None
+
+
+@register_augmented_forward(PrimIDs.SYNCHRONIZE_TENSOR_PARALLEL_OUTPUT)
+def synchronize_tensor_parallel_output_forward_rule(
+    t: TensorProxy,
+    group: torch.distributed.ProcessGroup,
+    layer_type: TensorParallelLayerType,
+) -> tuple[TensorProxy, tuple[torch.distributed.ProcessGroup, TensorParallelLayerType]]:
+    from thunder.distributed.tensor_parallel.common import TensorParallelLayerType
+    import thunder.torch as ltorch
+
+    match layer_type:
+        case TensorParallelLayerType.COLUMN_PARALLEL_LINEAR:
+            # all-gather in the last dim
+            future_of_all_gathered = all_gather(t, group, True, 0)
+            all_gathered = wait(future_of_all_gathered)
+            chunked = ltorch.chunk(all_gathered, group.size(), 0)
+            gathered = ltorch.cat(chunked, dim=-1)
+            return gathered, (group, layer_type)
+        case TensorParallelLayerType.ROW_PARALLEL_LINEAR:
+            return all_reduce(t, DistributedReduceOps.SUM, group, do_async=True, skip_clone=True).wait(), (
+                group,
+                layer_type,
+            )
+        case TensorParallelLayerType.COLUMN_PARALLEL_EMBED:
+            return all_reduce(t, DistributedReduceOps.SUM, group, do_async=True, skip_clone=True).wait(), (
+                group,
+                layer_type,
+            )
+        case TensorParallelLayerType.ROW_PARALLEL_EMBED:
+            # all-gather in the last dim
+            future_of_all_gathered = all_gather(t, group, True, 0)
+            all_gathered = wait(future_of_all_gathered)
+            chunked = ltorch.chunk(all_gathered, group.size(), 0)
+            gathered = ltorch.cat(chunked, dim=-1)
+            return gathered, (group, layer_type)
+        case _:
+            utils.check(False, lambda: f"Invalid {layer_type=}")
+
+
+@register_backward(PrimIDs.SYNCHRONIZE_TENSOR_PARALLEL_OUTPUT)
+def synchronize_tensor_parallel_output_backward_rule(
+    group: torch.distributed.ProcessGroup,
+    layer_type: TensorParallelLayerType,
+    grad: TensorProxy,
+) -> tuple[TensorProxy, None, None]:
+    from thunder.distributed.tensor_parallel.common import TensorParallelLayerType
+
+    match layer_type:
+        case TensorParallelLayerType.COLUMN_PARALLEL_LINEAR:
+            return (
+                reduce_scatter(
+                    grad / group.size(), DistributedReduceOps.SUM, group, do_async=True, dim=grad.ndim - 1
+                ).wait(),
+                None,
+                None,
+            )
+        case TensorParallelLayerType.ROW_PARALLEL_LINEAR:
+            return grad, None, None
```

Comment:
Thinking about it more: I know nothing about it, but if AllReduce produces 4 copies of the output, should the backward sum the for gradients corresponding to it? (based on the picture as https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allreduce )